packages/core/src/libraries/domain.ts

```diff
@@ -0,0 +1,105 @@
+import {
+  type CloudflareData,
+  type Domain,
+  type DomainDnsRecords,
+  DomainStatus,
+} from '@logto/schemas';
+
+import type Queries from '#src/tenants/Queries.js';
+import SystemContext from '#src/tenants/SystemContext.js';
+import assertThat from '#src/utils/assert-that.js';
+import { getCustomHostname, createCustomHostname } from '#src/utils/cloudflare/index.js';
+
+export type DomainLibrary = ReturnType<typeof createDomainLibrary>;
+
+const getDomainStatusFromCloudflareData = (data: CloudflareData): DomainStatus => {
+  if (data.status === 'active' && data.ssl.status === 'active') {
+    return DomainStatus.Active;
+  }
+
+  if (data.status === 'active' && data.ssl.status !== 'active') {
+    return DomainStatus.PendingSsl;
+  }
+
+  if (data.status !== 'active') {
+    return DomainStatus.PendingVerification;
+  }
+
+  return DomainStatus.Error;
+};
+
+export const createDomainLibrary = (queries: Queries) => {
+  const {
+    domains: { updateDomainById },
+  } = queries;
+
+  const syncDomainStatusFromCloudflareData = async (
+    domain: Domain,
+    cloudflareData: CloudflareData,
+    origin: string
+  ): Promise<Domain> => {
+    const status = getDomainStatusFromCloudflareData(cloudflareData);
+
+    const verificationError = cloudflareData.verification_errors?.join('\n');
+    const sslError = cloudflareData.ssl.validation_errors?.map(({ message }) => message).join('\n');
+    const errorMessage = [verificationError, sslError].filter(Boolean).join('\n');
+
+    const ownershipRecord = cloudflareData.ownership_verification;
+    const { txt_name, txt_value } = cloudflareData.ssl;
+    const sslRecord =
+      txt_name && txt_value ? { type: 'TXT', name: txt_name, value: txt_value } : undefined;
+    const cnameRecord =
+      status === DomainStatus.PendingVerification || status === DomainStatus.Error
+        ? {
+            type: 'CNAME',
+            name: domain.domain,
+            value: origin,
+          }
+        : undefined;
```

Comment:
`conditional`? (we can shorten this name in essentials though, like `c`, `cond`)